File: snake-game/python/tests/snake_builder.py
from snake_game.position import Position
from snake_game.direction import Direction
from snake_game.snake import Snake


class SnakeBuilder:
    def __init__(self) -> None:
        self._body: list[Position] = [Position(0, 0)]
        self._direction: Direction = Direction.RIGHT

    def at(self, x: int, y: int) -> "SnakeBuilder":
        self._body = [Position(x, y)]
        return self

    def with_body_at(self, *positions: tuple[int, int]) -> "SnakeBuilder":
        self._body = [Position(x, y) for x, y in positions]
        return self

    def moving_up(self) -> "SnakeBuilder":
        self._direction = Direction.UP
        return self

    def moving_down(self) -> "SnakeBuilder":
        self._direction = Direction.DOWN
        return self

    def moving_left(self) -> "SnakeBuilder":
        self._direction = Direction.LEFT
        return self

    def moving_right(self) -> "SnakeBuilder":
        self._direction = Direction.RIGHT
        return self

    def build(self) -> Snake:
        return Snake(self._body, self._direction)

File: snake-game/typescript/tests/SnakeBuilder.ts
import type { Position } from '../src/Position.js';
import type { Direction } from '../src/Direction.js';
import { Snake } from '../src/Snake.js';

export class SnakeBuilder {
  private _body: Position[] = [{ x: 0, y: 0 }];
  private _direction: Direction = 'Right';

  at(x: number, y: number): this {
    this._body = [{ x, y }];
    return this;
  }

  withBodyAt(...positions: [number, number][]): this {
    this._body = positions.map(([x, y]) => ({ x, y }));
    return this;
  }

  movingUp(): this { this._direction = 'Up'; return this; }
  movingDown(): this { this._direction = 'Down'; return this; }
  movingLeft(): this { this._direction = 'Left'; return this; }
  movingRight(): this { this._direction = 'Right'; return this; }

  build(): Snake {
    return new Snake(this._body, this._direction);
  }
}
